File: Examples/android/CMeSdkJava/app_rework/src/main/java/me/connect/sdk/java/sample/backups/BackupsViewModel.java
package me.connect.sdk.java.sample.backups;

import android.app.Application;

import androidx.lifecycle.AndroidViewModel;

import java.util.concurrent.Executors;

import me.connect.sdk.java.WalletBackups;
import me.connect.sdk.java.sample.Constants;
import me.connect.sdk.java.sample.SingleLiveData;
import me.connect.sdk.java.sample.db.Database;
import me.connect.sdk.java.sample.db.entity.Backup;

public class BackupsViewModel extends AndroidViewModel {
    private final Database db;


    public BackupsViewModel(Application application) {
        super(application);
        db = Database.getInstance(application);
    }

    public SingleLiveData<String> performBackup() {
        SingleLiveData<String> data = new SingleLiveData<>();
        backup(data);
        return data;
    }

    private void backup(SingleLiveData<String> liveData) {
        Executors.newSingleThreadExecutor().execute(() -> {
            WalletBackups.create(getApplication(), Constants.WALLET_NAME, "secret_key", "backup").handle((res, err) -> {
                if (res != null) {
                    Backup backup = new Backup();
                    backup.id = 1;
                    backup.path = res;
                    db.backupDao().insertAll(backup);
                    liveData.postValue("Saved to: " + res);
                } else {
                    err.printStackTrace();
                    liveData.postValue("Error: " + err.toString());
                }
                return null;
            });
        });
    }

    public SingleLiveData<String> performRestore() {
        SingleLiveData<String> data = new SingleLiveData<>();
        restore(data);
        return data;
    }

    private void restore(SingleLiveData<String> liveData) {
        Executors.newSingleThreadExecutor().execute(() -> {
            Backup backup = db.backupDao().getById(1);
            WalletBackups.restore(getApplication(), "secret_key", backup.path).handle((res, err) -> {
                if (err == null) {
                    liveData.postValue("Success");
                } else {
                    err.printStackTrace();
                    liveData.postValue("Error: " + err.toString());
                }
                return null;
            });
        });
    }

    public SingleLiveData<String> getLastBackup() {
        SingleLiveData<String> data = new SingleLiveData<>();
        getState(data);
        return data;
    }

    private void getState(SingleLiveData<String> liveData) {
        Executors.newSingleThreadExecutor().execute(() -> {
            Backup backup = db.backupDao().getById(1);
            if (backup != null) {
                liveData.postValue(backup.path);
            } else {
                liveData.postValue(null);
            }
        });
    }
}

File: Examples/android/CMeSdkJava/app_rework_kt/src/main/kotlin/me/connect/sdk/java/samplekt/backups/BackupsViewModel.kt
package me.connect.sdk.java.samplekt.backups

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import me.connect.sdk.java.WalletBackups
import me.connect.sdk.java.samplekt.Constants
import me.connect.sdk.java.samplekt.SingleLiveData
import me.connect.sdk.java.samplekt.db.Database
import me.connect.sdk.java.samplekt.db.entity.Backup


class BackupsViewModel(application: Application) : AndroidViewModel(application) {
    private val db = Database.getInstance(application)

    fun performBackup(): SingleLiveData<String> {
        val data = SingleLiveData<String>()
        backup(data)
        return data
    }

    private fun backup(liveData: SingleLiveData<String>) = viewModelScope.launch(Dispatchers.IO) {
        try {
            val res = WalletBackups.create(getApplication(), Constants.WALLET_NAME, "secret_key", "backup").get()
            val backup = Backup(
                    id = 1,
                    path = res
            )
            db.backupDao().insertAll(backup)
            liveData.postValue("Saved to: $res")
        } catch (e: Exception) {
            e.printStackTrace()
            liveData.postValue("Error: $e")
        }
    }

    fun performRestore(): SingleLiveData<String> {
        val data = SingleLiveData<String>()
        restore(data)
        return data
    }

    private fun restore(liveData: SingleLiveData<String>) = viewModelScope.launch(Dispatchers.IO) {
        try {
            val backup = db.backupDao().getById(1)!!
            WalletBackups.restore(getApplication(), "secret_key", backup.path).get()
            liveData.postValue("Success")
        } catch (e: Exception) {
            e.printStackTrace()
            liveData.postValue("Error: $e")
        }
    }

    fun getLastBackup(): SingleLiveData<String?> {
        val data = SingleLiveData<String?>()
        getState(data)
        return data
    }

    private fun getState(liveData: SingleLiveData<String?>) = viewModelScope.launch(Dispatchers.IO) {
        val backup = db.backupDao().getById(1)
        liveData.postValue(backup?.path)
    }

}
